tutorials/xr/openxr_composition_layers.rst

```diff
@@ -0,0 +1,293 @@
+
+                # Place our pointer where we're pointing
+                var pos : Vector3 = _intersect_to_global_pos(intersect)
+                $Pointer.visible = true
+                $Pointer.global_position = pos
+
+    ...
+
+If we were intersecting in our previous process call and our pointer has moved,
+we prepare a :ref:`InputEventMouseMotion <class_InputEventMouseMotion>` object
+to simulate our mouse moving and send that to our viewport for further processing.
+
+.. code:: gdscript
+
+    ...
+
+                if was_intersect != NO_INTERSECTION and intersect != was_intersect:
+                    # Pointer moved
+                    var event : InputEventMouseMotion = InputEventMouseMotion.new()
+                    var from : Vector2 = _intersect_to_viewport_pos(was_intersect)
+                    var to : Vector2 = _intersect_to_viewport_pos(intersect)
+                    if was_pressed:
+                        event.button_mask = MOUSE_BUTTON_MASK_LEFT 
+                    event.relative = to - from
+                    event.position = to
+                    layer_viewport.push_input(event)
+
+    ...
+
+If we've just released our button we also prepare
+a :ref:`InputEventMouseButton <class_InputEventMouseButton>` object
+to simulate a button release and send that to our viewport for further processing.
+
+.. code:: gdscript
+
+    ...
+
+                if !is_pressed and was_pressed:
+                    # Button was let go?
+                    var event : InputEventMouseButton = InputEventMouseButton.new()
+                    event.button_index = 1
+                    event.pressed = false
+                    event.position = _intersect_to_viewport_pos(intersect)
+                    layer_viewport.push_input(event)
+
+    ...
+
+Or if we've just pressed our button we prepare
+a :ref:`InputEventMouseButton <class_InputEventMouseButton>` object
+to simulate a button press and send that to our viewport for further processing.
+
+.. code:: gdscript
+
+    ...
+
+                elif is_pressed and !was_pressed:
+                    # Button was pressed?
+                    var event : InputEventMouseButton = InputEventMouseButton.new()
+                    event.button_index = 1
+                    event.button_mask = MOUSE_BUTTON_MASK_LEFT
+                    event.pressed = true
+                    event.position = _intersect_to_viewport_pos(intersect)
+                    layer_viewport.push_input(event)
+
+    ...
+
+Next we remember our state for next frame.
+
+.. code:: gdscript
+
+    ...
+
+                was_pressed = is_pressed
+                was_intersect = intersect
+
+    ...
+
+Finally, if we aren't intersecting, we simply clear our state.
+
+.. code:: gdscript
+
+    ...
+
+            else:
+                was_pressed = false
+                was_intersect = NO_INTERSECTION
+
+
+Hole punching
+-------------
+
+This is a feature that is specifically targeted at AR use cases as it requires our
```

Comment:
I'd add `MR` in addition to `AR` in that sentence to avoid giving the impression we're only referring to AR glasses:

> This is a feature that is specifically targeted at AR and MR use cases...

Reply:
As mentioned during the XR meeting, should have written this down before :) In my humble opinion I want to avoid the MR terminology here. AR and VR reference two distinct approaches, VR referencing fully immersive virtual worlds, while AR references any approach where the real world is augmented with virtual content.

Just for completeness sake in for this discussion, XR is a grouping term that was adopted by Khronos and as such by a part of the industry (including by Godot).

MR is a confusing term that means something different depending on who you talk to.
- Originally it referred to mixing real world capture with virtual world capture to create a composite such as solutions as MixCast, Liv and Oculus own Mixed Reality iPad app implement.
- Then Microsoft adopted MR to mean Mixed Reality as the name for their WMR platform, in this case defining it as a platform supported both AR (through Hololens) and VR (through other WMR devices, none of which do passthrough I believe). I also guess they hoped people would see it as an abbreviation of Microsoft Reality, but that's pure speculation on my part :)
- Then in most recent years MR meaning Mixed Reality has become an alternative to XR as an encompassing term for people who don't like the XR moniker.
- And finally, Meta seems to have rebranded it recently to specifically mean AR through the means of passthrough though I am not sure if that has created more clarification or more confusion in the market space. All though Metas marketing here IMHO is inconsistent as its also often referring to the above meaning as a catch all for AR and VR.

So MR in its meaning in the context of this document as I understand it, e.g. AR through passthrough, is a meaning that is fairly contested.

As Godot is a platform agnostic development tool and we try not to care **how** AR is achieved but want our applications to be portable between various AR devices, it is my personal opinion that AR in this context is sufficient capturing all use cases of this feature, while MR is confusing.

There are features that are only possible on devices that use a camera feed to achieve AR and where passthrough should be named specifically, but this is no such feature as it is as applicable on a passthrough device such as the Quest, as it is on an optical AR device such as the Magic Leap 2.

**However**, I stress that this is my personal standpoint on the matter, so this is absolutely up for debate.